fix tree agglomeration in agglomerateByPrevalence

theotimeee opened the pull request:
As mentioned in https://github.com/microbiome/mia/issues/573, agglomerateByPrevalence returns a TSE object with `rowTree = NULL`.

This PR removes the class checking and class conversion from TSE to SCE that removes the tree.

With this modification, the tree is returned and agglomerated if `agglomerate.tree = TRUE`.

```
data("GlobalPatterns")
tse <- GlobalPatterns
tse <- agglomerateByPrevalence(tse, rank = "Phylum", agglomerate.tree = FALSE)
rowTree(tse)
```
>Phylogenetic tree with 19216 tips and 19215 internal nodes.
>Tip labels:
  549322, 522457, 951, 244423, 586076, 246140, ...
Node labels:
  , 0.858.4, 1.000.154, 0.764.3, 0.995.2, 1.000.2, ...
>Rooted; includes branch lengths.

```
data("GlobalPatterns")
tse <- GlobalPatterns
tse <- agglomerateByPrevalence(tse, rank = "Phylum", agglomerate.tree = TRUE)
rowTree(tse)
```
>Phylogenetic tree with 67 tips and 66 internal nodes.
Tip labels:
  549322, 173903, 368907, 51888, 591769, 249529, ...
Node labels:
  , 0.858.4, 1.000.21479, 0.946.619, 0.953.588, 0.940.473, ...
Rooted; includes branch lengths.

Commit:
remove class conversion in agglomerateByPrevalence

TuomasBorman committed:
Merge branch 'devel' into fix_tree_prevalence

TuomasBorman reviewed (changes requested):
Can you add unit tests?

I found also that the tree does not correspond to rows of TreeSE. There are more tips in tree than there are rows. This is because the agglomeration is done in agglomerateByRank and then after rank agglomeration, agglomeratio is done by prevalence. However, user would expect that the agglomeration of tree is also done by prevalence. --> add tree agglomeration also in agglomerateByPrevalence

TuomasBorman committed:
Merge branch 'devel' into fix_tree_prevalence

Commit:
agglomerate tree by prevalence and add unit tests

theotimeee committed:
Merge branch 'devel' into fix_tree_prevalence

TuomasBorman commented on R/getPrevalence.R:
```diff
@@ -580,18 +586,11 @@ setMethod("agglomerateByPrevalence", signature = c(x = "SummarizedExperiment"),
-            x <- as(x,class_x)
+            x <- rbind(x[f,], other_x)
+        }
+        if (agglomerate.tree == TRUE){
```
`if( agglomerate.tree )`

TuomasBorman replied:
This was not addressed. As agglomerate.tree is already boolean, it is good practice to use it directly instead of comparing boolean value to boolean value

TuomasBorman reviewed (changes requested):
Minor fix.

Add also news and version bump

TuomasBorman commented:
Also add input check for agglomerate.tree. Fail fast.

Commit:
bump version, input check and minor fix

Commit:
Merge branch 'fix_tree_prevalence' of https://github.com/microbiome/mia into fix_tree_prevalence

TuomasBorman committed:
Merge branch 'devel' into fix_tree_prevalence

Commit:
fix if statement

Commit:
Merge branch 'fix_tree_prevalence' of https://github.com/microbiome/mia into fix_tree_prevalence

TuomasBorman commented:
Realized one thing: the function might not work if the input is SummarizedExperiment since it has no trees.

--> Create generic method for TreeSummarizedExperiment
1. Call next method which will can generic SE method
2. Merge trees
3. Merge reference sequences (see agglomerateByVariable)

Commit:
SE method for agglomerateByPrevalence and add merge for reference sequences

Commit:
fix documentation

TuomasBorman commented on R/getPrevalence.R:
```diff
@@ -580,19 +590,42 @@ setMethod("agglomerateByPrevalence", signature = c(x = "SummarizedExperiment"),
+        refSeq <- referenceSeq(x)
+      }
+      #
+      x <- .merge_rows(x, f, archetype = 1L, ...)
```
Wondering what is the .merge_rows doing here.

I think this line should be the callNextMethod() call

TuomasBorman replied:
This merges rows with rownames --> this does not do anything?

theotimeee replied:
Yes I did not notice that, I removed it but I already call callNextMethod() at the beginning of the method, should I move it to this line?

TuomasBorman replied:
I did not check in detail, but in theory reference sequence might already be modified when calling callNextMethod(). That is because the SE method then calls other agglomeration methods.

That is why it might be safest to first take the reference sequences and then call callNextMethod() so that you have stored the original reference sequences.

Moreover, I realized that these all agglomeration functions are in same documentation page. agglomerateByVariable() can be done also in column-wise which means that it must have mergeRefSeq parameter as "hidden" parameter. --> agglomerateByPrevalence() should also have the parameter as hidden (in "...")

TuomasBorman replied:
You could do it with `list(...)[["mergeRefSeq"]]` without creating an internal function. Just explain in code comment, why this is done that way

TuomasBorman commented on R/getPrevalence.R:
```diff
@@ -580,19 +590,42 @@ setMethod("agglomerateByPrevalence", signature = c(x = "SummarizedExperiment"),
+      if(!is.null(refSeq)){
+        referenceSeq(x) <- .merge_refseq_list(refSeq, f, rownames(x), ...)
+      }
+      x
```
Use return(x)

Commit:
remove .merge_rows call

TuomasBorman commented on R/getPrevalence.R:
```diff
@@ -517,7 +517,17 @@ setMethod("getPrevalentAbundance", signature = c(x = "SummarizedExperiment"),
 #'   summary of non-prevalent taxa. (default: \code{other_label = "Other"})
-#'
+#'   
+#' @param agglomerate.tree \code{TRUE} or \code{FALSE}: should
+#'   \code{rowTree()} also be agglomerated? (Default:
+#'   \code{agglomerate.tree = FALSE})
+#'
+#' @param mergeRefSeq logical scalar: Should a consensus sequence be calculated?
+#' If set to \code{FALSE}, the result
+#' from \code{archetype} is returned; If set to \code{TRUE} the result
+#' from \code{\link[DECIPHER:ConsensusSequence]{DECIPHER::ConsensusSequence}}
+#' is returned. (Default: \code{mergeRefSeq = FALSE})
```
These will be changed by this PR https://github.com/microbiome/mia/pull/582 which takes the priority

Commit:
merge reference sequences before calling callNextMethod()

TuomasBorman commented on R/getPrevalence.R:
```diff
@@ -601,30 +595,34 @@ setMethod("agglomerateByPrevalence", signature = c(x = "SummarizedExperiment"),
+      # mergeRefSeq is a hidden parameter as for all other agglomeration methods
+      # from the agglomerate-methods man page.
+      # Here 'list(...)[["mergeRefSeq]]' is used to access it.
+      if( "mergeRefSeq" %in% names(list(...)) && 
+          !(is.null(list(...)[["mergeRefSeq"]]) || 
+            .is_a_bool(list(...)[["mergeRefSeq"]])) ){
+        stop("'mergeRefSeq' must be TRUE or FALSE.", call. = FALSE)
```
To make this more readable, get the value first to variable and then use it

TuomasBorman reviewed (changes requested):
Looks good

Commit:
make code more readable

TuomasBorman commented on R/getPrevalence.R:
```diff
@@ -580,19 +584,46 @@ setMethod("agglomerateByPrevalence", signature = c(x = "SummarizedExperiment"),
+      # from the agglomerate-methods man page.
+      # Here 'list(...)[["mergeRefSeq]]' is used to access it.
+      mergeRefSeq <- list(...)[["mergeRefSeq"]]
+      if( "mergeRefSeq" %in% names(list(...)) && 
+          !(is.null(mergeRefSeq) || 
+            .is_a_bool(mergeRefSeq)) ){
```
It is enough to test

```
merge_refseq <- list(...)[["update.refseq"]]
!is.null(merge_refseq) && !.is_a_bool(merge_refseq)
```

If that parameter is not available the value is NULL. If the value is not NULL and not boolean, it should lead to error.

TuomasBorman replied:
Let's get back to drawing board, this is too complex.

```
merge_refseq <- list(...)[["update.refseq"]]
if( is.null(merge_refseq) {
merge_refseq <- FALSE
}
if( !.is_a_bool(merge_refseq) ){
error
}
if( merge_refseq ){
refSeq <- referenceSeq(x)
}
....
...
...

if( merge_refseq ){
referenceSeq(x) <- .merge_refseq_list(refSeq, f, rownames(x), ...)
}
```

TuomasBorman commented on R/getPrevalence.R:
```diff
@@ -580,19 +584,46 @@ setMethod("agglomerateByPrevalence", signature = c(x = "SummarizedExperiment"),
+            .is_a_bool(mergeRefSeq)) ){
+        stop("'mergeRefSeq' must be TRUE or FALSE.", call. = FALSE)
+      }
+      else if( "mergeRefSeq" %in% names(list(...)) ){
+        refSeq <- referenceSeq(x)
+      }
```
Then this could be

if( !is.null(merge_refseq) && merge_refseq ){
    refSeq <- referenceSeq(x)
}

TuomasBorman replied:
This comment is outdated.

Commit:
merge devel and make review changes

Merge branch 'devel' of https://github.com/microbiome/mia into fix_tree_prevalence

# Conflicts:
#	NEWS
#	R/getPrevalence.R
#	tests/testthat/test-5prevalence.R

Commit:
up

TuomasBorman commented on R/getPrevalence.R:
```diff
@@ -590,19 +594,46 @@ setMethod("agglomerateByPrevalence", signature = c(x = "SummarizedExperiment"),
+      if( merge_refseq ){
+        refSeq <- referenceSeq(x)
+      }
+      if( merge_refseq ){
+        referenceSeq(x) <- .merge_refseq_list(refSeq, f, rownames(x), ...)
+      }
+      x <- callNextMethod()
```
You should probably first call

x <- callNextMethod()

and then

 if( merge_refseq ){
        referenceSeq(x) <- .merge_refseq_list(refSeq, f, rownames(x), ...)
 }
 
 I have not tested the code yet, but it looks like it

Commit:
callNextMethod() before merging ref sequences

TuomasBorman committed:
up

TuomasBorman commented:
Agglomeration of reference sequences was not working, but now everything should be fine.

Can you add tests for reference sequences? Check other agglomeration functions and their unit tests.

You can find suitable data from data("SilvermanAGutData", package = "miaTime")

Commit:
up doc

Commit:
add tests for reference sequences with agglomerateByPrevalence

Commit:
fix merge seq test

TuomasBorman commented on R/getPrevalence.R:
```diff
@@ -638,7 +638,7 @@ setMethod("agglomerateByPrevalence",
         # Update tree if user has specified to do so
         if( update.tree ){
-            res <- .agglomerate_trees(res, 1)
+          res <- .agglomerate_trees(res, 1)
```
Indentation off

TuomasBorman commented on tests/testthat/test-5prevalence.R:
```diff
@@ -452,7 +452,7 @@ test_that("agglomerateByPrevalence", {
       threshold = th)
     seqs_test <- seqs_test[ names(seqs_test) %in% feature ]
-    expect_equal(seqs_test, seqs_ref)
+    expect_equal(seqs_test, seqs_ref, tolerance = 10**-2)
```
Explain with the comment why there is tolerance (the tolerance is also quite high)

theotimeee replied:
Actually after modification tolerance is not needed

Commit:
update.tree instead of agglomerate.tree in test file

Commit:
fix identation

Commit:
remove tolerance and check merge of refseq with prevalence agglomeration on phylum

Commit:
merge devel

Merge branch 'devel' of https://github.com/microbiome/mia into fix_tree_prevalence

# Conflicts:
#	NEWS

Commit:
test on Genus with agglomerateByPrevalence works

TuomasBorman reviewed (approved):
Cool!
